server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -881,4 +900,91 @@ public void onNonemptyQueue() {
         }
     }
 
+    private static class TimingStatisticsTracker {
```

Comment:
All methods `synchronized` (but very fast) just to avoid getting partially-updated stats.